# Итоговый проект по блоку "Работа аналитика в компании"

In [1]:
%config IPCompleter.greedy = True
import pandas as pd
import numpy as np
pd.options.display.max_rows = 999

In [2]:
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
plt.style.use('ggplot')
plt.rcParams['figure.figsize'] = (12,8)

In [3]:
df = pd.read_excel('data_sample.xlsx', index_col = 'InvoiceDate')
df.head()

InvoiceNo StockCode                     Description  \
InvoiceDate                                                               
2010-12-01 08:26:00    536365    84029E  RED WOOLLY HOTTIE WHITE HEART.   
2010-12-01 08:34:00    536367     84879   ASSORTED COLOUR BIRD ORNAMENT   
2010-12-01 08:34:00    536367     21755        LOVE BUILDING BLOCK WORD   
2010-12-01 08:34:00    536368     22914    BLUE COAT RACK PARIS FASHION   
2010-12-01 09:02:00    536373    84406B  CREAM CUPID HEARTS COAT HANGER   

                     Quantity  UnitPrice  CustomerID         Country  
InvoiceDate                                                           
2010-12-01 08:26:00         6       3.39     17850.0  United Kingdom  
2010-12-01 08:34:00        32       1.69     13047.0  United Kingdom  
2010-12-01 08:34:00         3       5.95     13047.0  United Kingdom  
2010-12-01 08:34:00         3       4.95     13047.0  United Kingdom  
2010-12-01 09:02:00         8       2.75     17850.0  United Kingdom

#### Краткое описание таблицы:
- InvoiceDate - дата заказа
- InvoiceNo - номер заказа
- StockCode - код товара
- Description - описание товара
- Quantity - количество товара
- UnitPrice - цена за единицу товара
- CustomerID - id клиента
- Country - страна

#### Первый взгляд на данные

In [4]:
df.shape

(54214, 7)

In [5]:
df.describe()

Quantity     UnitPrice    CustomerID
count  54214.000000  54214.000000  40643.000000
mean       9.610802      4.307341  15283.422803
std       57.358422     47.834496   1707.982278
min    -9600.000000      0.000000  12347.000000
25%        1.000000      1.250000  13969.000000
50%        3.000000      2.080000  15152.000000
75%       10.000000      4.130000  16775.000000
max     2400.000000   6930.000000  18287.000000

In [6]:
df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

In [7]:
df.isnull().sum()

InvoiceNo          0
StockCode          0
Description      134
Quantity           0
UnitPrice          0
CustomerID     13571
Country            0
dtype: int64

In [8]:
df.duplicated().sum()

50

In [9]:
df[df.duplicated() == True]

InvoiceNo StockCode                          Description  \
InvoiceDate                                                                    
2010-12-05 13:55:00    537196     22640       SET OF 4 NAPKIN CHARMS 3 KEYS    
2010-12-06 11:26:00    537265     22960             JAM MAKING SET WITH JARS   
2010-12-06 12:41:00    537369     47471       RAINBOW PEGS PHOTO CLIP STRING   
2010-12-06 14:31:00    537399     22815              CARD PSYCHEDELIC APPLES   
2010-12-07 15:39:00    537646     22189              CREAM HEART CARD HOLDER   
2010-12-07 15:39:00    537646     21484          CHICK GREY HOT WATER BOTTLE   
2010-12-07 15:54:00    537653     22107                   PIZZA PLATE IN BOX   
2010-12-08 12:46:00    537781     21625             VINTAGE UNION JACK APRON   
2010-12-08 12:46:00    537781    72799E     IVORY PILLAR CANDLE SILVER FLOCK   
2010-12-12 14:27:00    538514     21756             BATH BUILDING BLOCK WORD   
2010-12-14 10:38:00    538707     20981        12 PENCILS TALL TUBE WOODLAND   
2010-12-14 13:31:00    538849     22073            RED RETROSPOT STORAGE JAR   
2011-01-09 12:53:00    540524     21326     AGED GLASS SILVER T-LIGHT HOLDER   
2011-01-16 16:23:00    541265    84459A            PINK METAL CHICKEN HEART    
2011-01-19 17:17:00    541602    85061W      WHITE JEWELLED HEART DECORATION   
2011-01-24 16:05:00    541987     22667                RECIPE BOX RETROSPOT    
2011-02-27 11:40:00    545045     84880                WHITE WIRE EGG HOLDER   
2011-03-30 12:02:00    548309     21034         REX CASH+CARRY JUMBO SHOPPER   
2011-05-12 20:03:00    553034     22103   MIRROR MOSAIC T-LIGHT HOLDER ROUND   
2011-05-18 11:28:00    553657     22060      LARGE CAKE STAND HANGING HEARTS   
2011-05-19 08:26:00    553731     23208        LUNCH BAG VINTAGE LEAF DESIGN   
2011-05-20 10:32:00    553935    35810A         ENAMEL PINK COFFEE CONTAINER   
2011-05-27 15:57:00    554964     22252  BIRDCAGE DECORATION TEALIGHT HOLDER   
2011-06-08 11:18:00    555979     22470                HEART OF WICKER LARGE   
2011-06-15 13:25:00    556910     22196         SMALL HEART MEASURING SPOONS   
2011-07-31 11:26:00    561868     22438    BALLOON ART MAKE YOUR OWN FLOWERS   
2011-08-05 15:38:00    562540     23073                 GEORGIAN TRINKET BOX   
2011-08-11 14:04:00    563026    85099F                 JUMBO BAG STRAWBERRY   
2011-08-24 12:04:00    564303     22553               PLASTERS IN TIN SKULLS   
2011-09-08 12:42:00    566018     20676                   RED RETROSPOT BOWL   
2011-09-28 15:42:00    568710     23399        HOME SWEET HOME HANGING HEART   
2011-10-07 15:45:00    570197     21031                    SPACE CADET BLACK   
2011-10-21 13:58:00   C572226     85066          CREAM SWEETHEART MINI CHEST   
2011-10-23 16:19:00    572318     22810            SET OF 6 T-LIGHTS SNOWMEN   
2011-10-26 12:39:00    572859     22795         SWEETHEART RECIPE BOOK STAND   
2011-10-30 11:55:00    573340     20719               WOODLAND CHARLOTTE BAG   
2011-11-03 16:30:00    574304     20979        36 PENCILS TUBE RED RETROSPOT   
2011-11-04 12:05:00    574474     85227       SET OF 6 3D KIT CARDS FOR KIDS   
2011-11-08 10:44:00    574965     23351             ROLL WRAP 50'S CHRISTMAS   
2011-11-09 14:11:00    575337     22030               SWALLOWS GREETING CARD   
2011-11-13 14:09:00    575985    85199S    SMALL HANGING IVORY/RED WOOD BIRD   
2011-11-13 15:30:00    576062     82482    WOODEN PICTURE FRAME WHITE FINISH   
2011-11-18 12:48:00    577297     22154          ANGEL DECORATION 3 BUTTONS    
2011-11-20 11:56:00    577485     22940            FELTCRAFT CHRISTMAS FAIRY   
2011-11-20 13:23:00    577522     23168         CLASSIC CAFE SUGAR DISPENSER   
2011-11-25 13:45:00    578817     21810     CHRISTMAS HANGING STAR WITH BELL   
2011-11-29 14:38:00    579458     82583                 HOT BATHS METAL SIGN   
2011-12-02 11:16:00    580164    84970S    HANGING HEART ZINC T-LIGHT HOLDER   
2011-12-04 11:00:00 

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 54214 entries, 2010-12-01 08:26:00 to 2011-12-09 12:50:00
Data columns (total 7 columns):
InvoiceNo      54214 non-null object
StockCode      54214 non-null object
Description    54080 non-null object
Quantity       54214 non-null int64
UnitPrice      54214 non-null float64
CustomerID     40643 non-null float64
Country        54214 non-null object
dtypes: float64(2), int64(1), object(4)
memory usage: 3.3+ MB


In [11]:
df.sort_values(by='Quantity').head(50)

InvoiceNo StockCode                          Description  \
InvoiceDate                                                                    
2011-06-14 10:37:00    556691     23005         printing smudges/thrown away   
2011-10-11 11:10:00   C570556     20971     PINK BLUE FELT CRAFT TRINKET BOX   
2011-10-31 15:18:00    573598    79342B               Unsaleable, destroyed.   
2011-10-31 15:18:00    573597     79341               Unsaleable, destroyed.   
2011-09-21 09:16:00   C567527     23113                PANTRY CHOPPING BOARD   
2010-12-09 14:48:00    538090     20956                                    ?   
2011-09-14 12:04:00    566712     22030                                   ??   
2011-01-31 12:36:00   C542693     15036            ASSORTED COLOURS SILK FAN   
2011-10-11 11:10:00   C570556     22149           FELTCRAFT 6 FLOWER FRIENDS   
2011-03-08 17:23:00    546018     85172                 wrongly sold as sets   
2011-03-23 17:27:00    547559     72759              thrown away-can't sell.   
2011-10-24 10:40:00   C572343     22103   MIRROR MOSAIC T-LIGHT HOLDER ROUND   
2011-10-03 14:19:00    569341     82600  historic computer difference?....se   
2011-08-31 13:10:00   C564940     22564               ALPHABET STENCIL CRAFT   
2011-05-23 10:40:00    554143    79030G                              damages   
2011-07-28 14:58:00    561639     84799                                  NaN   
2011-12-08 09:56:00    581226     23090                              missing   
2011-12-07 18:34:00    581207     21688                             mixed up   
2011-08-31 17:02:00   C565044     22191               IVORY DINER WALL CLOCK   
2011-01-12 15:04:00    540978     84050                              counted   
2011-09-23 14:50:00   C568096    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER   
2011-11-28 15:43:00   C579192     22959               WRAP CHRISTMAS VILLAGE   
2011-08-31 17:21:00    565071     84754                                  NaN   
2010-12-14 13:27:00   C538847    85232B      SET OF 3 BABUSHKA STACKING TINS   
2011-05-10 12:06:00   C552569         D                             Discount   
2011-02-07 12:10:00    543310     35954                                  NaN   
2011-08-31 13:10:00   C564940     23203             JUMBO BAG VINTAGE DOILY    
2011-11-21 10:18:00   C577614     21499                   BLUE POLKADOT WRAP   
2011-04-01 10:46:00   C548577     21155                RED RETROSPOT PEG BAG   
2011-02-25 16:26:00   C545033     22630                 DOLLY GIRL LUNCH BOX   
2011-08-31 17:02:00   C565044     11001          ASSTD DESIGN RACING CAR PEN   
2011-03-11 16:24:00    546407     22719                wrong barcode (22467)   
2011-08-18 10:43:00    563650     79163                                  NaN   
2010-12-09 13:38:00    538058     20966                                  NaN   
2011-10-13 12:15:00    571012     21402                                  NaN   
2010-12-10 16:01:00   C538357     35924        HANGING FAIRY CAKE DECORATION   
2011-08-31 17:02:00   C565044     21479        WHITE SKULL HOT WATER BOTTLE    
2010-12-02 16:58:00   C536812     22130     PARTY CONE CHRISTMAS DECORATION    
2011-11-30 14:36:00    579743     21810                                check   
2011-08-31 17:02:00   C565044     21731        RED TOADSTOOL LED NIGHT LIGHT   
2011-06-15 16:54:00    556943    84809A                                  NaN   
2011-10-06 10:57:00   C569743     23199                     JUMBO BAG APPLES   
2011-01-04 11:10:00    540009     22158                                  NaN   
2011-12-06 13:41:00   C580913     22536           MAGIC DRAWING SLATE PURDEY   
2011-09-14 17:51:00    566766     21311                                  NaN   
2011-07-08 10:36:00    559373    16156S                                  NaN   
2011-03-24 16:39:00    547700     21147                                  NaN   
2011-10-24 10:25:00   C572334     22386              JUMBO BAG PINK POLKADOT   
2011-11-09 11:41:00 

In [12]:
df[df.Country == 'Unspecified']

InvoiceNo StockCode                          Description  \
InvoiceDate                                                                    
2011-04-11 13:29:00    549687     20685                DOORMAT RED RETROSPOT   
2011-04-11 13:29:00    549687     21212      PACK OF 72 RETROSPOT CAKE CASES   
2011-05-10 15:31:00    552695     22619            SET OF 6 SOLDIER SKITTLES   
2011-05-10 15:31:00    552695     23127            FELTCRAFT GIRL NICOLE KIT   
2011-05-10 15:31:00    552695     22585            PACK OF 6 BIRDY GIFT TAGS   
2011-05-10 15:31:00    552695     22993         SET OF 4 PANTRY JELLY MOULDS   
2011-05-10 15:31:00    552695     22485        SET OF 2 WOODEN MARKET CRATES   
2011-05-10 15:31:00    552695     82482    WOODEN PICTURE FRAME WHITE FINISH   
2011-07-08 16:26:00    559521     21634        ASSORTED MINI MADRAS NOTEBOOK   
2011-07-08 16:26:00    559521     22317         FIVE CATS HANGING DECORATION   
2011-07-08 16:26:00    559521     22434        BALLOON PUMP WITH 10 BALLOONS   
2011-07-08 16:26:00    559521     22439                   6 ROCKET BALLOONS    
2011-07-14 09:10:00    559929     22485        SET OF 2 WOODEN MARKET CRATES   
2011-07-14 09:10:00    559929    46000M         POLYESTER FILLER PAD 45x45cm   
2011-07-28 16:06:00    561658     84813        SET OF 4 DIAMOND NAPKIN RINGS   
2011-07-28 16:06:00    561658     23014       GLASS APOTHECARY BOTTLE ELIXIR   
2011-07-28 16:06:00    561658     21592         RETROSPOT CIGAR BOX MATCHES    
2011-07-28 16:06:00    561658     21843             RED RETROSPOT CAKE STAND   
2011-07-28 16:06:00    561658     23202               JUMBO BAG VINTAGE LEAF   
2011-07-28 16:06:00    561658     23012      GLASS APOTHECARY BOTTLE PERFUME   
2011-07-28 16:21:00    561661    79063C    RETRO PILL BOX KEY CHAIN,THE KING   
2011-07-28 16:21:00    561661     21260                        FIRST AID TIN   
2011-08-22 10:18:00    563947     23236           STORAGE TIN VINTAGE DOILY    
2011-08-22 10:18:00    563947     23318       BOX OF 6 MINI VINTAGE CRACKERS   
2011-08-22 13:32:00    564051     22845     VINTAGE CREAM CAT FOOD CONTAINER   
2011-09-02 12:17:00    565303     22747           POPPY'S PLAYHOUSE BATHROOM   
2011-09-02 12:17:00    565303     23309  SET OF 60 I LOVE LONDON CAKE CASES    
2011-09-02 12:17:00    565303     22499            WOODEN UNION JACK BUNTING   
2011-09-02 12:17:00    565303     22964   3 PIECE SPACEBOY COOKIE CUTTER SET   
2011-09-02 12:17:00    565303     21558        SKULL LUNCH BOX WITH CUTLERY    
2011-11-16 10:18:00    576646     22554     PLASTERS IN TIN WOODLAND ANIMALS   
2011-11-16 10:18:00    576646     22556       PLASTERS IN TIN CIRCUS PARADE    
2011-11-24 14:55:00    578539     22748            POPPY'S PLAYHOUSE KITCHEN   
2011-11-24 14:55:00    578539     22559                  SEASIDE FLYING DISC   
2011-11-24 14:55:00    578539     23507     MINI PLAYING CARDS BUFFALO BILL    
2011-11-24 14:55:00    578539     23570     TRADITIONAL PICK UP STICKS GAME    

                     Quantity  UnitPrice  CustomerID      Country  
InvoiceDate                                                        
2011-04-11 13:29:00         2       7.95     12363.0  Unspecified  
2011-04-11 13:29:00        24       0.55     12363.0  Unspecified  
2011-05-10 15:31:00         4       3.75     16320.0  Unspecified  
2011-05-10 15:31:00         4       4.95     16320.0  Unspecified  
2011-05-10 15:31:00        12       1.25     16320.0  Unspecified  
2011-05-10 15:31:00        12       1.25     16320.0  Unspecified  
2011-05-10 15:31:00         2      12.75     16320.0  Unspecified  
2011-05-10 15:31:00         6       2.55     16320.0  Unspecified  
2011-07-08 16:26:00        24       0.39         NaN  Unspecified  
2011-07-08 16:26:00         1       2.95         NaN  Unspecified  
2011-07-08 16:26:00         6       1.95         NaN  Unspecified  
2011-07-08 16:26:00        40       0.65         NaN  Unspecified  
2011-07-14 09:10:00        

In [13]:
df[df.StockCode == 'D']

InvoiceNo StockCode Description  Quantity  UnitPrice  \
InvoiceDate                                                                
2010-12-08 16:00:00   C537857         D    Discount        -1     267.12   
2011-02-11 17:10:00   C543789         D    Discount        -1      84.80   
2011-03-03 11:05:00   C545476         D    Discount        -1      30.38   
2011-05-10 12:06:00   C552569         D    Discount      -240       0.03   
2011-08-09 14:27:00   C562796         D    Discount        -1      64.00   
2011-08-18 06:13:00   C563593         D    Discount        -1     206.40   
2011-08-31 12:29:00   C564909         D    Discount        -1      33.75   
2011-08-31 17:25:00   C565075         D    Discount        -1      36.30   
2011-09-13 17:17:00   C566613         D    Discount        -1      15.00   
2011-11-03 12:39:00   C574238         D    Discount        -1      11.84   
2011-11-11 12:52:00   C575872         D    Discount        -1      97.14   
2011-11-14 13:40:00   C576244         D    Discount        -1       6.00   
2011-11-18 12:06:00   C577227         D    Discount        -1      19.82   

                     CustomerID         Country  
InvoiceDate                                      
2010-12-08 16:00:00     17340.0  United Kingdom  
2011-02-11 17:10:00     17450.0  United Kingdom  
2011-03-03 11:05:00     14527.0  United Kingdom  
2011-05-10 12:06:00     12901.0  United Kingdom  
2011-08-09 14:27:00     13694.0  United Kingdom  
2011-08-18 06:13:00     14646.0     Netherlands  
2011-08-31 12:29:00     15498.0  United Kingdom  
2011-08-31 17:25:00     13408.0  United Kingdom  
2011-09-13 17:17:00     16029.0  United Kingdom  
2011-11-03 12:39:00     13408.0  United Kingdom  
2011-11-11 12:52:00     13408.0  United Kingdom  
2011-11-14 13:40:00     18167.0  United Kingdom  
2011-11-18 12:06:00     14527.0  United Kingdom

In [14]:
df[df['Quantity'] < 0].count().Quantity

1083

In [15]:
df[(df['UnitPrice'] < 0) & (df['Quantity'] < 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, UnitPrice, CustomerID, Country]
Index: []

##### В таблице обнаружены отрицательные и неизвестные значения, дубликаты.
- От дубликатов лучше всего избавиться.
- Имеется StockCode 'D', отвечающий за скидку. Поскольку изначально значения неизвестны, считать отрицательным его количество не будем, чтобы значения не добавлялись к остальным, где были факты упущенной выгоды
- Отрицательные цены и количество товаров, судя по описанию некоторых товаров, отражают потерянные, списанные товары
- К сожалению, не во всех местах, где количество товаров отрицательное, указана цена, чтобы оценить экономический масштаб потерь
- К сожалению, в некоторых местах неизвестна страна, однако мы допустим, что товар был доставлен, а выручка получена
- Также много пустых значений id клиента - мы их оставим и представим, что они просто неизвестны, а прочие манипуляции были выполнены

In [16]:
df['CustomerID'] = df.CustomerID.fillna(-1).astype(int)
df.loc[df['CustomerID'] == -1, 'CustomerID'] = 'Unknown'

In [17]:
df.drop_duplicates(inplace=True)
df.shape

(54164, 7)

In [18]:
df.loc[df.StockCode == 'D', 'Quantity'] = 0

#### Вычленяем информацию из данных

In [19]:
df['TotalPrice'] = df['UnitPrice'] * df['Quantity']

In [20]:
df['Weekday'] = df.index.weekday.map({0: 'Monday', 1: 'Tuesday', 2: 'Wednesday', 3: 'Thursday', 4: 'Friday', 5: 'Saturday', 6: 'Sunday'})

In [21]:
df['Month'] =df.index.month

In [22]:
df['Year'] = df.index.year

In [23]:
df['Day'] = df.index.day

In [24]:
df.Country.value_counts()

United Kingdom          49553
Germany                   950
France                    881
EIRE                      787
Spain                     253
Netherlands               234
Belgium                   211
Switzerland               189
Portugal                  152
Australia                 130
Norway                    105
Channel Islands            81
Italy                      77
Finland                    72
Cyprus                     61
Sweden                     50
Austria                    48
Poland                     44
Unspecified                36
Denmark                    35
Japan                      33
Hong Kong                  29
Israel                     27
Singapore                  25
USA                        23
Iceland                    15
Greece                     14
Canada                     13
Malta                       9
Lithuania                   6
European Community          5
United Arab Emirates        5
RSA                         4
Brazil    

#### Что мы можем сказать на данный момент?
- Это некая компания, которая, по всей видимости, базируется в Великобритании.
- Основная нагрузка лежит также на Великобританию, однако небольшая часть товаров все же отправляется за границу.

#### Посмотрим, на какую сумму компания понесла убытки

In [25]:
df[df['Country'] == 'European Community']

InvoiceNo StockCode                          Description  \
InvoiceDate                                                                    
2011-04-26 10:54:00    551013     22841         ROUND CAKE TIN VINTAGE GREEN   
2011-07-21 10:24:00    560783     22449             SILK PURSE BABUSHKA PINK   
2011-07-21 10:24:00    560783     23020          GLASS  SONGBIRD STORAGE JAR   
2011-07-21 10:24:00    560783     23268  SET OF 2 CERAMIC CHRISTMAS REINDEER   
2011-07-21 10:24:00    560783     21098                CHRISTMAS TOILET ROLL   

                     Quantity  UnitPrice CustomerID             Country  \
InvoiceDate                                                               
2011-04-26 10:54:00         2       7.95      15108  European Community   
2011-07-21 10:24:00         6       3.35      15108  European Community   
2011-07-21 10:24:00         1      12.50      15108  European Community   
2011-07-21 10:24:00        12       1.45      15108  European Community   
2011-07-21 10:24:00        12       1.25      15108  European Community   

                     TotalPrice   Weekday  Month  Year  Day  
InvoiceDate                                                  
2011-04-26 10:54:00        15.9   Tuesday      4  2011   26  
2011-07-21 10:24:00        20.1  Thursday      7  2011   21  
2011-07-21 10:24:00        12.5  Thursday      7  2011   21  
2011-07-21 10:24:00        17.4  Thursday      7  2011   21  
2011-07-21 10:24:00        15.0  Thursday      7  2011   21

In [26]:
lost_revenue = abs(df[df['TotalPrice'] < 0].groupby('Country')['Quantity', 'TotalPrice'].sum()).reset_index()
lost_revenue.style.bar()

In [27]:
df.astype(str).sort_values(by='InvoiceNo', ascending=False).reset_index().head(200)

InvoiceDate InvoiceNo  StockCode  \
0   2011-12-08 18:59:00   C581465      22171   
1   2011-12-08 11:57:00   C581330      22959   
2   2011-12-08 11:51:00   C581322      22666   
3   2011-12-08 10:14:00   C581229      21507   
4   2011-12-08 10:14:00   C581229     79191C   
5   2011-12-08 10:06:00   C581228      22781   
6   2011-12-07 15:39:00   C581178      23079   
7   2011-12-07 13:48:00   C581145      23250   
8   2011-12-07 12:24:00   C581117      22988   
9   2011-12-06 14:57:00   C580968      23526   
10  2011-12-06 14:19:00   C580954     84997B   
11  2011-12-06 14:19:00   C580954      22128   
12  2011-12-06 14:07:00   C580919      23240   
13  2011-12-06 14:07:00   C580919      23493   
14  2011-12-06 13:41:00   C580913      22536   
15  2011-12-06 13:37:00   C580912      22041   
16  2011-12-06 11:40:00   C580832      23480   
17  2011-12-06 11:40:00   C580832      23238   
18  2011-12-06 10:54:00   C580768      23020   
19  2011-12-06 10:32:00   C580762      22960   
20  2011-12-06 10:31:00   C580761      22139   
21  2011-12-06 10:31:00   C580761      22943   
22  2011-12-06 09:56:00   C580751      23201   
23  2011-12-05 16:36:00   C580711      85066   
24  2011-12-05 16:33:00   C580708      84946   
25  2011-12-05 15:28:00   C580686      22358   
26  2011-12-05 15:28:00   C580686      22488   
27  2011-12-02 12:43:00   C580263          M   
28  2011-12-01 15:06:00   C580099      23485   
29  2011-12-01 13:12:00   C580056      22467   
30  2011-12-01 10:56:00   C579948      21730   
31  2011-12-01 10:41:00   C579945      22197   
32  2011-12-01 09:29:00   C579928      23485   
33  2011-12-01 08:46:00   C579906      22621   
34  2011-12-01 08:29:00   C579897      22470   
35  2011-12-01 08:26:00   C579894      23568   
36  2011-12-01 08:18:00   C579891      23485   
37  2011-12-01 08:14:00   C579890      84947   
38  2011-11-30 17:42:00   C579887      84946   
39  2011-11-30 17:39:00   C579886      22594   
40  2011-11-30 15:20:00   C579781      22456   
41  2011-11-30 15:14:00   C579778      23460   
42  2011-11-30 14:56:00   C579757      47469   
43  2011-11-30 14:41:00   C579748      23229   
44  2011-11-30 09:21:00   C579532      22890   
45  2011-11-28 15:43:00   C579192      22959   
46  2011-11-28 14:55:00   C579178      22941   
47  2011-11-25 15:18:00   C578832      23286   
48  2011-11-25 15:18:00   C578832      22469   
49  2011-11-25 15:18:00   C578832      84987   
50  2011-11-25 15:18:00   C578832      22379   
51  2011-11-25 15:18:00   C578832      22867   
52  2011-11-25 12:55:00   C578803      22191   
53  2011-11-24 11:16:00   C578371      22551   
54  2011-11-24 11:16:00   C578371      22555   
55  2011-11-23 13:48:00   C578276      22197   
56  2011-11-23 13:48:00   C578276      22665   
57  2011-11-23 10:30:00   C578124      22659   
58  2011-11-23 10:30:00   C578124      21561   
59  2011-11-22 17:31:00   C578097      23355   
60  2011-11-22 16:02:00   C578073          M   
61  2011-11-22 16:02:00   C578073          M   
62  2011-11-22 11:42:00   C577889      22248   
63  2011-11-22 10:59:00   C577849      22910   
64  2011-11-22 10:18:00   C577832      21231   
65  2011-11-22 10:18:00   C577832      22989   
66  2011-11-22 09:53:00   C577828      23318   
67  2011-11-22 09:29:00   C577821     72802A   
68  2011-11-21 16:55:00   C577799      23130   
69  2011-11-21 13:25:00   C577737      23022   
70  2011-11-21 13:25:00   C577737      22599   
71  2011-11-21 13:25:00   C577737      23221   
72  2011-11-21 13:25:00   C577737      22734   
73  2011-11-21 11:47:00   C577692      22798   
74  2011-11-21 10:51:00   C577683      22726   
75  2011-11-21 10:18:00   C577614      21499   
76  2011-11-20 12:46:00   C577509      21846   
77  2011-11-18 17:01:00   C577390      23401   
78  2011-11-18 16:54:00   C577386      23401   
79  2011-11-18 16:24:00   C577381      23012   
80  2011-11-18 15:16:00   C577345      23243   
81  2011-11-18 12:06:00   C577227          D   
82  2011-11-18 11:45:00   C5

In [28]:
df[df.UnitPrice == 0]

InvoiceNo     StockCode  \
InvoiceDate                                   
2010-12-01 14:35:00    536553         37461   
2010-12-03 15:29:00    536995         35951   
2010-12-03 15:30:00    536998         85067   
2010-12-03 15:35:00    537005         21692   
2010-12-03 15:38:00    537009        84534B   
2010-12-03 15:42:00    537018        84562A   
2010-12-07 11:48:00    537534         85064   
2010-12-07 11:48:00    537534         22681   
2010-12-07 11:48:00    537534         22520   
2010-12-07 11:48:00    537534         22371   
2010-12-07 11:48:00    537534         21912   
2010-12-07 11:48:00    537534         20615   
2010-12-08 17:08:00    537870         21486   
2010-12-08 18:08:00    537876         84977   
2010-12-09 13:38:00    538058         20966   
2010-12-09 14:09:00    538071         22585   
2010-12-09 14:48:00    538090         20956   
2010-12-09 15:54:00    538128         21357   
2010-12-09 15:55:00    538130        90042C   
2010-12-09 15:56:00    538134         20878   
2010-12-09 17:17:00    538158         20892   
2010-12-14 14:04:00    538861         22444   
2010-12-16 14:36:00    539263         22580   
2010-12-20 14:33:00    539611        85135B   
2010-12-21 15:40:00    539750         21481   
2010-12-22 14:41:00    539856         22072   
2010-12-22 14:41:00    539856         20719   
2011-01-04 11:10:00    540008        85230F   
2011-01-04 11:10:00    540009         22158   
2011-01-10 16:05:00    540674         22837   
2011-01-10 16:20:00    540677         21667   
2011-01-12 15:04:00    540978         84050   
2011-01-12 15:35:00    540982         22323   
2011-01-19 09:11:00    541519         21429   
2011-01-27 13:59:00    542380         84917   
2011-01-28 14:52:00    542561        84424A   
2011-02-01 13:07:00    542883         21448   
2011-02-04 17:25:00    543270         84631   
2011-02-07 12:10:00    543310         35954   
2011-02-10 10:05:00    543552         22847   
2011-02-28 14:19:00    545176         22360   
2011-02-28 14:19:00    545176         22064   
2011-02-28 14:19:00    545176         22383   
2011-02-28 14:19:00    545176        84968E   
2011-02-28 14:19:00    545176        84872A   
2011-02-28 14:19:00    545176         37333   
2011-02-28 14:19:00    545176         22697   
2011-02-28 14:19:00    545176         22699   
2011-02-28 14:19:00    545176         22698   
2011-03-03 16:12:00    545560         37464   
2011-03-08 17:23:00    546018         85172   
2011-03-09 16:36:00    546141        79302M   
2011-03-11 16:21:00    546406        46000S   
2011-03-11 16:24:00    546407         22719   
2011-03-18 11:02:00    546933         22623   
2011-03-18 11:02:00    546933         22372   
2011-03-22 11:37:00    547328         22272   
2011-03-22 11:56:00    547346         21000   
2011-03-22 11:57:00    547348        85206A   
2011-03-23 17:27:00    547559         72759   
2011-03-24 16:39:00    547700         21147   
2011-03-24 17:51:00    547704         21621   
2011-03-29 10:18:00    547994         21228   
2011-03-29 12:10:00    548069         20713   
2011-03-29 17:21:00    548209         21621   
2011-03-30 16:00:00    548369         84683   
2011-03-30 16:46:00    548382        84898F   
2011-04-01 12:34:00    548638         21258   
2011-04-05 13:19:00    548989        72799F   
2011-04-05 14:34:00    548999        84743C   
2011-04-05 14:39:00    549009         21474   
2011-04-06 15:22:00    549133         20759   
2011-04-06 15:22:00    549134         20758   
2011-04-06 17:13:00    549157         21511   
2011-04-06 17:15:00    549161         21518   
2011-04-06 17:33:00    549178         47502   
2011-04-07 12:45:00    549266         84593   
2011-04-08 10:45:00    549326         21738   
2011-04-08 10:46:00    549327        84912A   
2011-04-08 10:55:00    549332        84864A   
2011-04-08 11:07:00    549343         72815   
2011-04-08 11:21:00    549355        84899F   
2011-04-08 11:27:00    549371         84921   
2011-04-08 14:20:00    54

In [29]:
df[(df.UnitPrice > 0) & df.CustomerID == 'Unknown']

C:\Program Files (x86)\Soft data analyst\Anaconda\lib\site-packages\pandas\core\ops\__init__.py:1115: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, UnitPrice, CustomerID, Country, TotalPrice, Weekday, Month, Year, Day]
Index: []

In [30]:
print('Как минимум компания не продала товара на {:.1f} у.е.'.format(abs(lost_revenue.TotalPrice.sum())))
print('В количестве это:', abs(lost_revenue.Quantity.sum()), 'штук')

Как минимум компания не продала товара на 55179.6 у.е.
В количестве это: 13452.0 штук


In [31]:
a = abs(df[df['TotalPrice'] < 0].sum().TotalPrice / df[df['TotalPrice'] > 0].sum().TotalPrice)
print('Это составляет {:.1f} % от всего дохода с продаж'.format(a *100))

Это составляет 5.2 % от всего дохода с продаж


#### Теперь посмотрим на доход компании

In [32]:
revenue = df[df['Quantity'] > 0].groupby('Country')['TotalPrice','Quantity'].sum().reset_index()[:-1]\
        .sort_values(by=['TotalPrice'], ascending=False)
revenue

Country  TotalPrice  Quantity
35        United Kingdom  892000.541    460266
10                  EIRE   29008.610     14305
24           Netherlands   26489.180     20930
14               Germany   23689.660     12208
13                France   19811.480     11423
0              Australia   17156.590     10009
32           Switzerland    6215.900      3541
30                 Spain    5971.220      2936
31                Sweden    4936.520      4440
3                Belgium    4066.660      2399
27              Portugal    4000.070      1831
25                Norway    3859.900      1800
6        Channel Islands    2877.950      1567
20                 Japan    2164.180      1534
9                Denmark    2058.840       742
12               Finland    2040.600      1017
19                 Italy    1475.470       763
7                 Cyprus    1057.950       851
1                Austria    1036.270       449
26                Poland     911.590       465
29             Singapore     883.080       373
16             Hong Kong     879.010       373
18                Israel     531.720       331
15                Greece     490.820       158
22             Lithuania     291.600       104
33                   USA     290.940       127
23                 Malta     287.900        42
17               Iceland     271.790       161
5                 Canada     214.140       226
34  United Arab Emirates      96.420        38
8         Czech Republic      82.200        36
11    European Community      80.900        33
4                 Brazil      79.800        28
28                   RSA      42.140        29
21               Lebanon      39.600         8
2                Bahrain      17.700         6

In [33]:
print('Доля дохода от экспорта за границу по отношению ко всему доходу {:.1f} %'.format(revenue.TotalPrice[1:].sum() / revenue.TotalPrice.sum() * 100))
print('По отношению в количестве это:', round(revenue.Quantity[1:].sum() / revenue.Quantity.sum() * 100, 1), '%')

Доля дохода от экспорта за границу по отношению ко всему доходу 15.5 %
По отношению в количестве это: 17.2 %


In [34]:
print('Суммарный доход равен: {:.0f} у.е.'.format(df.TotalPrice.sum()))

Суммарный доход равен: 1000750 у.е.


In [35]:
sales_by_month = df.groupby(['Year', 'Month'])['Quantity', 'TotalPrice'].sum()
sales_by_month.style.bar()

In [36]:
df['weekday_index'] = df.index.weekday
sales_by_weekday = df.groupby(['weekday_index', 'Weekday'])['Quantity', 'TotalPrice'].sum()
#df.drop('weekday_index', axis=1, inplace=True)
sales_by_weekday.style.bar()

In [37]:
df[df['Country'] != 'United Kingdom'].groupby(['weekday_index', 'Weekday'])['Quantity', 'TotalPrice'].sum().style.bar()

In [38]:
df[df['Country'] == 'United Kingdom'].groupby(['weekday_index', 'Weekday'])['Quantity', 'TotalPrice'].sum().style.bar()

In [39]:
df[df['Country'] != 'United Kingdom'].groupby(['Year', 'Month'])['Quantity', 'TotalPrice'].sum().style.bar()

In [40]:
np.where(df.groupby(['Year', 'Month', 'Day']).count()['InvoiceNo'] > 250)[0]

array([  0,   3,   4,   5,   6,   7,   8,  14,  31,  86, 109, 217, 236,
       237, 242, 247, 249, 250, 252, 258, 259, 264, 269, 270, 272, 274,
       275, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288,
       289, 290, 291, 292, 294, 295, 296, 297, 298, 300, 301, 303],
      dtype=int64)

In [45]:
df.to_excel('AIC_final.xlsx')

### Более подробное описание компании:
- Судя по названиям товаров, это онлайн-магазин подарков, который базируется в Великобритании.



###### Описание значений
- товар имеет цену 0 - товар утерян, не туда отправлен, сломан и т.д.
- количество товара отрицательное - возврат товара (кодировка C в начале)
- возможно, там, где неизвестен id клиента, строки использовались как пометки и изменения для конкретного товара в базе

###### Информация по доходам
- Примерный доход за год оказался 1 миллион у.е.
- При этом 15% дохода от продажи оказалось от экспорта

###### Информация по убыткам
- более 5% по отношению ко всему доходу (как минимум!) было утеряно, списано, отправлено не туда, возвращено

###### Сезонность
- по месяцам прослеживается сезонность: больше заказов становится непосредственно перед праздниками. Самые глобальные - новогодние праздники. Повышение продаж начинается месяца за полтора-два до них с расчетом на доставку.
- анализ продаж по дням недели показывает, что продаж вообще нет в субботу. Возможно, заказы отражаются, когда их принимают работники, а суббота - выходной день. Видимое преимущество перед другими днями имеет четверг и следом соседние дни.

###### Прочее
- К сожалению, данные у нас предоставлены только за год, поэтому проследить закономерность по отношению к прочим годам мы не можем

- Предположительное название магазина - Rex London (https://www.rexlondon.com), ранее назывался DotComGiftShop